### PR TITLE
Update Treasury to Support Relay Chain Block Number Provider

### PR DESCRIPTION
**Pull request by shawntabrizi**
The goal of this PR is to have the treasury pallet work on a parachain which does not produce blocks on a regular schedule, thus can use the relay chain as a block provider.

Because blocks are not produced regularly, we cannot make the assumption that block number increases monotonically, and thus have new logic to handle multiple spend periods passing between blocks.

### CODE REVIEW

File: substrate/frame/treasury/src/lib.rs [hunk 1]
```diff
@@ -451,17 +462,27 @@ pub mod pallet {
 			}
 
 			// Check to see if we should spend some funds!
-			if (n % T::SpendPeriod::get()).is_zero() {
-				Self::spend_funds()
+			let last_spend_period =
+				// AUDIT REVIEW TODO! Needs to handle migration story for this better.
+				LastSpendPeriod::<T, I>::get().unwrap_or(BlockNumberFor::<T>::zero());
```
Comment (shawntabrizi) [inline substrate/frame/treasury/src/lib.rs:467]: call out here

probably needs migration, but perhaps there is a way to avoid it with some clever logic?

**Review by bkchr — DISMISSED** [review]

File: substrate/frame/treasury/src/lib.rs (hunk 1 above)
Comment (xlc) [reply substrate/frame/treasury/src/lib.rs:468]:
```suggestion
				LastSpendPeriod::<T, I>::get().unwrap_or_else(|| calculate_last_spend_period());
```

in that way no migration is needed if the block number provider is the local block number

but we will need a migration when switch the block number provider

Comment (shawntabrizi) [reply substrate/frame/treasury/src/lib.rs:468]: > but we will need a migration when switch the block number provider

Seems unavoidable

File: substrate/frame/treasury/src/lib.rs [hunk 2]
```diff
@@ -437,7 +448,8 @@ pub mod pallet {
 	impl<T: Config<I>, I: 'static> Hooks<BlockNumberFor<T>> for Pallet<T, I> {
 		/// ## Complexity
 		/// - `O(A)` where `A` is the number of approvals
-		fn on_initialize(n: frame_system::pallet_prelude::BlockNumberFor<T>) -> Weight {
+		fn on_initialize(_do_not_use_local_block_number: BlockNumberFor<T>) -> Weight {
+			let block_number = T::BlockNumberProvider::current_block_number();
```
Comment (shawntabrizi) [inline substrate/frame/treasury/src/lib.rs:452]: This might be a common error when people write pallets depending on a `BlockNumberProvider`

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5831372

File: substrate/frame/treasury/src/lib.rs [hunk 3]
```diff
@@ -289,6 +293,9 @@ pub mod pallet {
 		/// Helper type for benchmarks.
 		#[cfg(feature = "runtime-benchmarks")]
 		type BenchmarkHelper: ArgumentsFactory<Self::AssetKind, Self::Beneficiary>;
+
+		/// Provider for the block number.
+		type BlockNumberProvider: BlockNumberProvider<BlockNumber = BlockNumberFor<Self>>;
```
Comment (muharem) [inline substrate/frame/treasury/src/lib.rs:298]: may be a comment that this is the system pallet in general case

**Review by muharem — APPROVED** [review]

File: substrate/frame/treasury/src/lib.rs [hunk 4]
```diff
@@ -289,6 +293,9 @@ pub mod pallet {
 		/// Helper type for benchmarks.
 		#[cfg(feature = "runtime-benchmarks")]
 		type BenchmarkHelper: ArgumentsFactory<Self::AssetKind, Self::Beneficiary>;
+
+		/// Provider for the block number.
```
Comment (shawntabrizi) [inline substrate/frame/treasury/src/lib.rs:297]:
```suggestion
		/// Provider for the block number. Normally this is the `frame_system` pallet.
```

File: substrate/frame/treasury/src/lib.rs (hunk 3 above)
Comment (shawntabrizi) [reply substrate/frame/treasury/src/lib.rs:298]: done

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5867947

File: substrate/frame/treasury/src/lib.rs [hunk 5]
```diff
@@ -453,17 +465,29 @@ pub mod pallet {
 			}
 
 			// Check to see if we should spend some funds!
-			if (n % T::SpendPeriod::get()).is_zero() {
-				Self::spend_funds()
+			let last_spend_period = LastSpendPeriod::<T, I>::get()
+				// This unwrap should only occur one time on any blockchain.
+				// `update_last_spend_period` will populate the `LastSpendPeriod` storage if it is
+				// empty.
+				.unwrap_or(Self::update_last_spend_period());
```
Comment (ggwpez) [inline substrate/frame/treasury/src/lib.rs:472]: It is possible to use the [`OnEmpty` syntax](https://github.com/paritytech/polkadot-sdk/blob/cd91c6b782b60906597ed28c4e4a4fdddd8428c9/substrate/frame/examples/kitchensink/src/lib.rs#L130-L137) of the `LastSpendPeriod` and then change it to `ValueQuery` - it is not really optional.

**Review by ggwpez — APPROVED** [review]

File: substrate/frame/treasury/src/lib.rs (hunk 5 above)
Comment (shawntabrizi) [reply substrate/frame/treasury/src/lib.rs:472]: Feel like it is exactly the same thing, but more substrate magic.

or what am I missing?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5888107

File: substrate/frame/treasury/src/lib.rs (hunk 5 above)
Comment (ggwpez) [reply substrate/frame/treasury/src/lib.rs:472]: Yup. Depends on whether you like more or less magic 😆

**Comment by ggwpez** [comment]
Hey @shawntabrizi, the benchmarks CI is still failing: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/6003177

**Comment by shawntabrizi** [comment]
@ggwpez i think it should be good now, but probably requires a more careful review as treasury changes leak into multiple other pallets

File: substrate/frame/child-bounties/src/benchmarking.rs [hunk 6]
```diff
@@ -114,7 +119,8 @@ fn activate_bounty<T: Config>(
 	let approve_origin =
 		T::SpendOrigin::try_successful_origin().map_err(|_| BenchmarkError::Weightless)?;
 	Bounties::<T>::approve_bounty(approve_origin, child_bounty_setup.bounty_id)?;
-	Treasury::<T>::on_initialize(BlockNumberFor::<T>::zero());
```
Comment (shawntabrizi) [inline substrate/frame/child-bounties/src/benchmarking.rs:117]: we shouldnt allow `on_initialize(0)`: https://github.com/paritytech/polkadot-sdk/issues/4316

File: prdoc/pr_3970.prdoc [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+# See doc at https://raw.githubusercontent.com/paritytech/polkadot-sdk/master/prdoc/schema_user.json
+
+title: Update Treasury to Support Block Number Provider
+
+doc:
+  - audience: Runtime Dev
+    description: |
+      The goal of this PR is to have the treasury pallet work on a parachain which does not produce blocks on a regular schedule, thus can use the relay chain as a block provider. Because blocks are not produced regularly, we cannot make the assumption that block number increases monotonically, and thus have new logic to handle multiple spend periods passing between blocks. To migrate existing treasury implementations, simply add `type BlockNumberProvider = System` to have the same behavior as before.
+
+crates: [ ]
```
Comment (muharem) [inline prdoc/pr_3970.prdoc:11]: we trying to be more smart with versioning now, here info - https://github.com/paritytech/polkadot-sdk/blob/master/docs/contributor/prdoc.md#record-semver-changes

Comment (shawntabrizi) [reply prdoc/pr_3970.prdoc:11]: can you make a suggestion for what should be put here?

File: substrate/frame/bounties/src/lib.rs [hunk 8]
```diff
@@ -551,8 +551,8 @@ pub mod pallet {
 						T::Currency::reserve(curator, deposit)?;
 						bounty.curator_deposit = deposit;
 
-						let update_due = frame_system::Pallet::<T>::block_number() +
-							T::BountyUpdatePeriod::get();
+						let update_due =
+							Self::treasury_block_number() + T::BountyUpdatePeriod::get();
```
Comment (ggwpez) [inline substrate/frame/bounties/src/lib.rs:555]: I imagine most of the logic breaks when a parachain migrates from `System` to the `Relay` `BlockNumberProvider`...  
Not sure what can be done about it without re-writing the pallet to use timepoints. Maybe we can use this as first step and advise to still use `System`.

File: substrate/frame/treasury/src/lib.rs (hunk 2 above)
Comment (ggwpez) [reply substrate/frame/treasury/src/lib.rs:454]: Yea, `on_initialize` should be deprecated in favour of the new hooks from https://github.com/paritytech/polkadot-sdk/pull/1781.

File: prdoc/pr_3970.prdoc (hunk 7 above)
Comment (muharem) [inline prdoc/pr_3970.prdoc:11]:
```suggestion
crates: 
- name: pallet-treasury
  bump: major
```

Comment (shawntabrizi) [reply prdoc/pr_3970.prdoc:11]: thank you

File: substrate/frame/bounties/src/lib.rs (hunk 8 above)
Comment (ggwpez) [reply substrate/frame/bounties/src/lib.rs:555]: I think in the long term we need some kind of block-number-abstraction like https://github.com/paritytech/polkadot-sdk/issues/3268, but its a bit stale now as low-prio. That is not necessarily timestamps, but could be implemented with Relay blocknumber or something else.

**Comment by muharem** [comment]
@shawntabrizi should we merge the PR? all required CIs are green

**Comment by shawntabrizi** [comment]
Its no problem from me. Perhaps requires an audit, but I leave it to you all to get this merged :)

I think the code makes sense, although adds yet another piece of complexity to pallet config.

**Comment by muharem** [comment]
@shawntabrizi do we have a live or upcoming use case for this?

**Comment by joepetrowski** [comment]
> @shawntabrizi do we have a live or upcoming use case for this?

Don't we need it for instances of Treasury on Collectives?

**Comment by muharem** [comment]
> > @shawntabrizi do we have a live or upcoming use case for this?
> 
> Don't we need it for instances of Treasury on Collectives?

I think we do not, Collectives chain does produce the blocks, and we do not need the spend period to be in sync with Relay Treasury spend period.

**Comment by shawntabrizi** [comment]
There isn't a world ending pressure to get this in, but I think this is a natural improvement to the treasury system for the new models of accessing and using coretime.

There was a project I was working on that would want this feature.

**Comment by bkchr** [comment]
> @shawntabrizi do we have a live or upcoming use case for this?

This question is totally irrelevant for this pr here. With block times getting more dynamic, we need this fix. There are also other downstream users than "we" that may use or want to use it.

File: substrate/frame/treasury/src/lib.rs (hunk 2 above)
Comment (kianenigma) [reply substrate/frame/treasury/src/lib.rs:429]: I am tempted to suggest removing the parameter from `Hooks` -- even without this change, it was superfluous as you (almost) always have access to block number from `frame_system` directly.

**Pull request merged**